From 5397bb6655ea8ec50d17d8d7e282558c86464285 Mon Sep 17 00:00:00 2001
From: Yue Tao <Yue.Tao@windriver.com>
Date: Mon, 18 Mar 2019 23:12:51 -0700
Subject: [PATCH] of: shadow the phandle cache before free it in
 __of_free_phandle_cache()

Upstream commit d7ba3e833 free phandle cache in a new function __of_free_phandle_cache(),
which conficts with the rt commit b75caf08, which free phandle cache outside of the devtree_lock.
To fix it, store the phandle cache in shadow before calling __of_free_phandle_cache().

drivers/of/base.c:120:2: error: 'shadow' undeclared (first use in this function)
  shadow = phandle_cache;
  ^~~~~~

Signed-off-by: Yue Tao <Yue.Tao@windriver.com>
---
 drivers/of/base.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/of/base.c b/drivers/of/base.c
index 904c768..a95905c 100644
--- a/drivers/of/base.c
+++ b/drivers/of/base.c
@@ -117,19 +117,21 @@ static void __of_free_phandle_cache(void)
 	for (k = 0; k < cache_entries; k++)
 		of_node_put(phandle_cache[k]);
 
-	shadow = phandle_cache;
 	phandle_cache = NULL;
 }
 
 int of_free_phandle_cache(void)
 {
 	unsigned long flags;
+	struct device_node **shadow;
 
 	raw_spin_lock_irqsave(&devtree_lock, flags);
 
+	shadow = phandle_cache;
 	__of_free_phandle_cache();
 
 	raw_spin_unlock_irqrestore(&devtree_lock, flags);
+	kfree(shadow);
 
 	return 0;
 }
@@ -169,6 +171,7 @@ void of_populate_phandle_cache(void)
 
 	raw_spin_lock_irqsave(&devtree_lock, flags);
 
+	shadow = phandle_cache;
 	__of_free_phandle_cache();
 
 	for_each_of_allnodes(np)
-- 
1.7.9.5

